# Battle of the Neighbourhoods - Segmenting and Clustering Neighbourhoods in Toronto

## IBM Data Science with Coursera - Capstone Project

This notebook is intended to be used for the Capstone project for Moudle 9 of the IBM Data Science with Coursera Course

## Table of Contents

* [1. Collecting and Cleaning Data](#collecting_and_cleaning_data)
 * [Importing Data](#importing_data)
 * [Cleaning Data](#cleaning_data)
* [2. Getting Location Data](#getting_location_data)
* [3. Clustering the Neighbourhoods](#clustering)
 * [Create the Toronto Map](#create_map)
 * [Find Location Data for our Neighbourhoods](#location_data)
 * [Analyze the Neighbourhoods](#analyze_neighbourhoods)
 * [Cluster the Neighbourhoods](#clustering)
 * [Mapping the Clusters](#mapping_clusters)

## 1. Collecting and Cleaning Data <a name="colelcting_and_cleaning_data"></a>

Install our dependencies

In [1]:
#install Beautiful Soup and requests for Web Scaping
!pip install BeautifulSoup4
!pip install requests

### Importing Data <a name="importing_data"></a>

In [2]:
wiki_page = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

Get HTML from wikipedia and assign to BeautifulSoup opject:

In [65]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import numpy as np

source_html = requests.get(wiki_page)
soup = BeautifulSoup(source_html.text, 'lxml')

Use BeautifulSoup to parse the html:

In [5]:
#using soup object, iterate the .wikitable to get the data from the HTML page and store it into a list
data = []     # the data for each postal code
columns = []  # for our table header
table = soup.find(class_='wikitable')
for index, tr in enumerate(table.find_all('tr')):
    section = []
    for td in tr.find_all(['th','td']):
        section.append(td.text.rstrip())
    
    #First row of data is the header
    if (index == 0):
        columns = section
    else:
        data.append(section)

Now turn this data into a pandas dataframe:

In [6]:
toronto_df = pd.DataFrame(data=data, columns=columns)
toronto_df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

### Cleaning Data <a name="cleaning_data"></a>

We can see that many of the postcodes are not assigned to a borough; we will remove these.

In [12]:
toronto_df = toronto_df[toronto_df['Borough'] != 'Not assigned']
toronto_df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Our index is now missing all the values that were removed, so we can reset our index:

In [25]:
toronto_df.reset_index(drop=True, inplace=True)
toronto_df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Finally, we check the shape of our dataframe:

In [26]:
toronto_df.shape

(103, 3)

## 2. Getting Location Data <a name="getting_location_data"></a>

We will use the more reliable geospatial data from http://cocl.us/Geospatial_data

In [31]:
geo_data_df = pd.read_csv('Geospatial_Coordinates.csv')
geo_data_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Now we can merge this location data with our existing toronto dataframe:

In [35]:
toronto_df = pd.merge(toronto_df, geo_data_df, on='Postal Code')
toronto_df.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## 3. Clustering the Neighbourhoods <a name=clustering></a>

As with New York, we first get the coordinates for Toronto

In [40]:
import folium

In [38]:
address = "Toronto, ON"

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto city are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto city are 43.6534817, -79.3839347.


### Create the Toronto Map <a name='create_map'></a>

We can create a map of all of Toronto with the neighbourhoods shown:

In [55]:
# create map of Toronto using latitude and longitude values
toronto_map = folium.Map(location=[latitude, longitude], zoom_start=10)

 Then we add the neighbourhood markers to the map:

In [56]:
for borough, neighbourhood, lat, lng in zip(
        toronto_df['Borough'], 
        toronto_df['Neighbourhood'],
        toronto_df['Latitude'], 
        toronto_df['Longitude'], 
        ):
    label = f'{neighbourhood}, {borough}'
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(toronto_map)  

toronto_map

### Find Location Data for our Neighbourhoods <a name="location_data"></a>

Now we get out foursquare credentials (to be removed before psoting to our public github repo!)

In [ ]:
CLIENT_ID = ''
CLIENT_SECRET = ''
VERSION = ''

Let's look at out first neighbourhood:


In [59]:
toronto_df.loc[0]

Postal Code             M3A
Borough          North York
Neighbourhood     Parkwoods
Latitude            43.7533
Longitude          -79.3297
Name: 0, dtype: object

So we can get the top 100 venues that are within 500m of Parkwoods:

In [61]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
neighborhood_latitude = toronto_df.loc[0, 'Latitude']
neighborhood_longitude = toronto_df.loc[0, 'Longitude']

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

# get the result to a json file
results = requests.get(url).json()

In [63]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [72]:
venues = results['response']['groups'][0]['items']
nearby_venues = pd.json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [column.split(".")[-1].capitalize() for column in nearby_venues.columns]

nearby_venues

Name         Categories        Lat        Lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

In [80]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    venues_list=[]
    
    for name, lat, lng in zip(names, latitudes, longitudes):         
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [81]:
toronto_venues = getNearbyVenues(names=toronto_df['Neighbourhood'],
                                 latitudes=toronto_df['Latitude'],
                                 longitudes=toronto_df['Longitude'],
                                 )

In [82]:
toronto_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

In [83]:
toronto_venues.shape

(2139, 7)

We can see that across Toronto we have found 2139 venues!

Let's check how many venues are in each neighbourhood:

In [85]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                             5   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Willowdale West                                           5   
Woburn                                                                4   
Woodbine Heights                                                      8   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                              5   
Alderwood, Long Branch                                                 7   
Bathurst Manor, Wilson Heights, Downsview North                       21   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     22   
...                                                                  ...   
Willowdale, Willowdale West                                            5   
Woburn                                                                 4   
Woodbine Heights                                                       8   
York Mills West                                                        2   
York Mills, Silver Hills                                               1   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            5               5   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Willowdale West                          5               5   
Woburn                                               4               4   
Woodbine Heights                                     8               8   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      5   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Willowdale West                                    5   
Woburn                                                         4   
Woodbine Heights                                               8   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     5  
Alderwood, Long Branch                

### Analyze the Neighbourhoods <a name='analyze_neighbourhoods'></a>

We use one-hot encoding to convert the venuw categories to numerical variables:

In [89]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighbourhood column back to dataframe
toronto_onehot.insert(loc=0, column='Neighbourhood', value=toronto_venues['Neighbourhood'])

toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2  Victoria Village                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  \
0                 0                    0  ...              0   
1                 0                    0  ...              0   
2                 0                    0  ...              0   
3                 0                    0  ...              0   
4                 0                    0  ...              0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                   0                              0                 0   
1                   0                              0                 0   
2                   0                              0                 0   
3                   0                              0                 0   
4                   0                              0                 0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 274 columns]

Now we group rows by neighbourhood:

In [93]:
toronto_onehot_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_onehot_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.045455   

   ...  Train Station  Turkish Restaurant  Vegetarian / Vegan Restaurant  \
0  ...            0.0                 0.0                            0.0   
1  ...            0.0                 0.0                            0.0   
2  ...            0.0                 0.0                            0.0   
3  ...            0.0                 0.0                            0.0   
4  ...            0.0                 0.0                            0.0   

   Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0               0.0                    0.0              0.0       0.0   
1               0.0                    0.0              0.0       0.0   
2               0.0                    0.0              0.0       0.0   
3               0.0                    0.0              0.0       0.0   
4               0.0                    0.0              0.0       0.0   

   Wings Joint  Women's Store  Yoga Studio  
0          0.0            0.0          0.0  
1          0.0            0.0          0.0  
2          0.0            0.0          0.0  
3          0.0            0.0          0.0  
4          0.0            0.0          0.0  

[5 rows x 274 columns]

We can check the 5 most common venue categories in each neighbourhood:

In [139]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except IndexError:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_onehot_grouped['Neighbourhood']

for ind in np.arange(toronto_onehot_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_onehot_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0          Skating Rink  Latin American Restaurant        Clothing Store   
1              Pharmacy                        Gym        Sandwich Place   
2           Coffee Shop        Fried Chicken Joint    Chinese Restaurant   
3                  Café         Chinese Restaurant                  Bank   
4    Italian Restaurant                Coffee Shop      Greek Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot    Falafel Restaurant           Event Space   
1           Coffee Shop                   Pub               Dog Run   
2           Bridal Shop        Sandwich Place                 Diner   
3   Distribution Center               Dog Run      Doner Restaurant   
4       Thai Restaurant             Locksmith          Liquor Store   

     8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0     Ethiopian Restaurant                Escape Room         Discount Store  
1       Dim Sum Restaurant                      Diner         Discount Store  
2               Restaurant  Middle Eastern Restaurant            Supermarket  
3               Donut Shop                  Drugstore            Yoga Studio  
4  Comfort Food Restaurant                  Juice Bar                Butcher

### Cluster the Neighbourhoods <a name="clustering"></a>

We use k-means clustering to cluster the neighbourhoods:

In [118]:
from sklearn.cluster import KMeans

In [130]:
clusters = 8
toronto_clustering = toronto_onehot_grouped.drop('Neighbourhood', axis=1)

k_means = KMeans(n_clusters=clusters, random_state=0).fit(toronto_clustering)

# check cluster labels generated for each row in the dataframe
k_means.labels_[0:100] 


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 6, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0,
       0, 0, 6, 0, 0, 5, 1, 0, 1, 0, 1, 0, 0, 0, 4, 0, 0, 6, 0, 0, 0, 1,
       0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0,
       6, 0, 0, 0, 0, 0, 6, 2])

In [140]:
# add clustering labels
neighbourhoods_venues_sorted.insert(1, 'Cluster', k_means.labels_)

In [141]:
neighbourhoods_venues_sorted.head()

Neighbourhood  Cluster  \
0                                        Agincourt        0   
1                           Alderwood, Long Branch        0   
2  Bathurst Manor, Wilson Heights, Downsview North        0   
3                                  Bayview Village        0   
4                Bedford Park, Lawrence Manor East        0   

  1st Most Common Venue 2nd Most Common Venue      3rd Most Common Venue  \
0                Lounge          Skating Rink  Latin American Restaurant   
1           Pizza Place              Pharmacy                        Gym   
2                  Bank           Coffee Shop        Fried Chicken Joint   
3   Japanese Restaurant                  Café         Chinese Restaurant   
4        Sandwich Place    Italian Restaurant                Coffee Shop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0        Clothing Store        Breakfast Spot    Falafel Restaurant   
1        Sandwich Place           Coffee Shop                   Pub   
2    Chinese Restaurant           Bridal Shop        Sandwich Place   
3                  Bank   Distribution Center               Dog Run   
4      Greek Restaurant       Thai Restaurant             Locksmith   

  7th Most Common Venue    8th Most Common Venue      9th Most Common Venue  \
0           Event Space     Ethiopian Restaurant                Escape Room   
1               Dog Run       Dim Sum Restaurant                      Diner   
2                 Diner               Restaurant  Middle Eastern Restaurant   
3      Doner Restaurant               Donut Shop                  Drugstore   
4          Liquor Store  Comfort Food Restaurant                  Juice Bar   

  10th Most Common Venue  
0         Discount Store  
1         Discount Store  
2            Supermarket  
3            Yoga Studio  
4                Butcher

We can add these cluster numbers to our neighbourhoods_venues_sorted dataframe:

In [145]:
toronto_clustered_neighbourhoods = toronto_df.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'),
                                                   on='Neighbourhood')
toronto_clustered_neighbourhoods['Cluster'] = toronto_clustered_neighbourhoods['Cluster'].fillna(0.0).astype(int)
toronto_clustered_neighbourhoods.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster 1st Most Common Venue 2nd Most Common Venue  \
0  43.753259 -79.329656        6                  Park     Food & Drink Shop   
1  43.725882 -79.315572        0          Intersection           Pizza Place   
2  43.654260 -79.360636        0           Coffee Shop                  Park   
3  43.718518 -79.464763        0        Clothing Store     Accessories Store   
4  43.662301 -79.389494        0           Coffee Shop           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0           Yoga Studio   Dumpling Restaurant     Distribution Center   
1          Hockey Arena           Coffee Shop   Portuguese Restaurant   
2                   Pub                Bakery                 Theater   
3              Boutique             Gift Shop  Furniture / Home Store   
4                  Bank              Beer Bar           Smoothie Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Dog Run      Doner Restaurant            Donut Shop   
1             Drugstore                 Diner        Discount Store   
2        Breakfast Spot                  Café        Farmers Market   
3           Event Space           Coffee Shop         Women's Store   
4        Sandwich Place            Restaurant                  Café   

   9th Most Common Venue 10th Most Common Venue  
0              Drugstore      Electronics Store  
1    Distribution Center                Dog Run  
2                    Spa             Beer Store  
3  Vietnamese Restaurant        Airport Service  
4     Chinese Restaurant  Portuguese Restaurant

### Mapping the Clusters <a name='mapping_clusters'></a>

In [123]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [148]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(clusters)
ys = [i + x + (i*x)**2 for i in range(clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(
        toronto_clustered_neighbourhoods['Latitude'], 
        toronto_clustered_neighbourhoods['Longitude'], 
        toronto_clustered_neighbourhoods['Neighbourhood'], 
        toronto_clustered_neighbourhoods['Cluster']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters